docs/content/users/extend/customizing-images.md

```diff
@@ -78,7 +78,7 @@ RUN chmod -R ugo+rw $COMPOSER_HOME
 ENV COMPOSER_HOME=""
 ```
 
-**Remember that the Dockerfile is building a Docker image that will be used later with DDEV.** At the time the Dockerfile is executing, your code is not mounted and the container is not running, it’s just being built. So for example, an `npm install` in `/var/www/html` will not do anything useful because the code is not there at image building time.
+**Remember that the Dockerfile is normally building a Docker image that will be used later with DDEV.** At the time the Dockerfile is executing, your code by default is not mounted and the container is not running, it’s just being built. So for example, an `npm install` in `/var/www/html` will not do anything useful because the code is not there at image building time. However, you use `RUN` in the context of your codebase using something like `RUN --mount=type=bind,source=.,target=/var/www/html cp /var/www/html/index.php /var/tmp/` would mount your code and make it available at the normal `/var/www/html` mount. However, you can't make changes to the mounted code, so things like an early `npm install` wouldn't work, although `npm install -g` would work.
```

Comment:
I’m having trouble following this.

It starts with one of those “remember” phrases that’s assuming I already know what you’re allegedly reminding me. If I’m being thick and the order of operations were already described somewhere, it’d be ideal to link to it. If this should have been obvious to the Docker-competent, fine.

Then a couple sentences that made sense are followed by two new “However,” clauses—and I’m not sure if the second however is countering the first or thickening the plot.

I ultimately don’t know what this is saying so I’m not sure what to recommend, other than writing it to explain like I’m five.

Reply:
And apologies in advance to any five-year-old Docker developers that would have read this just fine.

Reply:
It's all that problem of how much people know or don't know, and how little of that you can fix in this context.

The "normal" situation is that in a Dockerfile you're building an image, like building an image of a machine. This may happen months or years before the code of the project is even imagined. In the case of DDEV, most of each image is built during DDEV development. And then when DDEV is used by developers, their code is mounted into that container that was built so long ago.

In the case of additional .ddev/web-build/Dockerfile.*, the time may be closer to their code's lifecycle, but not necessarily. For example, they may be a `ddev start` for the first time that builds their extra stuff on May 1 and that same Docker layer may be in use unchanged when they do a `ddev start` on July 1. Because the Dockerfile content hasn't changed, no new build happens. And the entire thing is not done in the context of code... unless you do the one little thing suggested here, which may offer the possibility of contextual read-only code.

Hope that helps, and if it does, can help communicate this stuff more clearly.

Reply:
Thanks for that! I think I’m following and that you’re reiterating the varying degrees of difference between when images are being built and their relationship to code that runs—sort of a cautionary warning for Docker adventurers since there are potential pitfalls.

The immediate problems, I think are the “remember” and the dueling “howevers” that are just hard to parse.

More broadly, it’d be ideal to diagram or otherwise illustrate that lifecycle separately for context. Not as a matter of “fixing” peoples’ knowledge (which suggests it’s broken or insufficient) but educating them and helping them understand something you already do. _Then_ the “remember” can be justified, once that thing is clearly explained and something you can link to. This is probably a bigger and longer-term thing to improve, and I’d be happy to help with that if/when there’s a good chance.

Reply:
Yeah, would love to find ways to communicate this. Maybe add to https://ddev.readthedocs.io/en/latest/users/usage/architecture/ and explain difference between image, container, and mounted code?

This class of problem is a little like helping people with getting homebrew right when they don't know how $PATH works in bash or cmd. You have no basis for the discussion, so all are flailing in the wilderness.

Reply:
> Maybe add to https://ddev.readthedocs.io/en/latest/users/usage/architecture/ and explain difference between image, container, and mounted code?

I think that‘s a step in the right direction, maybe isolated in a context/page that’s less introductory and more of a deep dive. Specifically not overwhelming to someone just getting curious about how things work, with freedom to explain and illustrate without pressure to keep it light.

While it’s kind of off-topic, I’d still like to depict Linux vs. macOS vs. Windows+WSL2 on that architecture page, since the difference (especially with WSL2) can be conceptually challenging. That’s for anyone, including beginners. The Docker composition is probably for more of a novice—or at least a more daring—DDEV user.

> so all are flailing in the wilderness.

😆 All the more reason to be a vigilant forest ranger! These big gaps are common, and trying to bridge them is hard work, but I think the result is something more welcoming and empowering for everybody trying to understand and use the thing.

Reply:
Mostly linux/macOS/Windows are almost the same, with a slight tech difference for Linux. But WSL2 is baffling to people for multiple reasons. 3 levels of computer!